src/Raven.Server/Documents/Handlers/DocumentReplicationHandler.cs

```diff
@@ -14,79 +18,99 @@
-                                    {
-                                        result = await webSocket.ReceiveAsync(buffer, Database.DatabaseShutdown);
-                                        parser.SetBuffer(buffer.Array, result.Count);
-                                    }
-                                    writer.FinalizeDocument();
-                                    var message = writer.CreateReader();
-                                    string messageTypeAsString;
-                                    if (!message.TryGet(Constants.MessageType, out messageTypeAsString))
-                                        throw new InvalidDataException(
-                                            $"Got websocket message without a type. Expected property with name {Constants.MessageType}, but found none.");
+			                        //TODO : refactor this to more efficient method,
+			                        //since we do not care about contents of the property at this stage,
+			                        //but only about it's existence
+			                        if (!message.TryGet(Constants.MessageType, out messageTypeAsString))
+				                        throw new InvalidDataException(
+					                        $"Got websocket message without a type. Expected property with name {Constants.MessageType}, but found none.");
 
-                                    HandleMessage(
-                                        messageTypeAsString,
-                                        message,
-                                        webSocketStream,
-                                        context,
-                                        srcDbId,
-                                        incomingReplication);
-                                }
-                            }
-                            catch (Exception e)
-                            {
-                                throw new InvalidOperationException($@"Failed to receive replication document batch. (Origin -> Database Id = {srcDbId}, Database Name = {srcDbName}, Origin URL = {srcUrl})", e);
-                            }
+			                        HandleMessage(
+				                        messageTypeAsString,
+				                        message,
+				                        webSocketStream,
+				                        context,
+				                        srcDbId,
+				                        incomingReplication);
+		                        }
+	                        }
+	                        catch (Exception e)
+	                        {
+								await webSocket.CloseAsync(WebSocketCloseStatus.InternalServerError,
+									"Exception was thrown, cannot continue replication. Check logs for details", Database.DatabaseShutdown);
+		                        var msg =
+			                        $@"Failed to receive replication document batch. (Origin -> Database Id = {srcDbId}, Database Name = {srcDbName}, Origin URL = {srcUrl})";
+								Log.ErrorException(msg,e);
```

Comment:
Log before the send, and you are throwing exception anywya, which also should be logged, shouldn't be logged twice.